pkg/services/accesscontrol/scope.go

```diff
@@ -44,33 +55,113 @@ func Field(key string) string {
 	return fmt.Sprintf(`{{ .%s }}`, key)
 }
 
+// ScopeMutator alters a Scope to return a new modified Scope
+type ScopeMutator func(context.Context, string) (string, error)
+
 type KeywordScopeResolveFunc func(*models.SignedInUser) (string, error)
 
-// ScopeResolver contains a map of functions to resolve scope keywords such as `self` or `current` into `id` based scopes
+// ScopeResolver is used to resolve scope keywords such as `self` or `current` into `id` based scopes and scope attributes such as `name` or `uid` into `id` based scopes.
 type ScopeResolver struct {
-	keywordResolvers map[string]KeywordScopeResolveFunc
+	keywordResolvers   map[string]KeywordScopeResolveFunc
+	attributeResolvers map[string]AttributeScopeResolveFunc
+	cache              *localcache.CacheService
+	log                log.Logger
 }
 
 func NewScopeResolver() ScopeResolver {
 	return ScopeResolver{
 		keywordResolvers: map[string]KeywordScopeResolveFunc{
 			"users:self": resolveUserSelf,
 		},
+		attributeResolvers: map[string]AttributeScopeResolveFunc{},
+		cache:              localcache.New(ttl, cleanInterval),
+		log:                log.New("accesscontrol.scoperesolution"),
 	}
 }
 
+func (s *ScopeResolver) AddKeywordResolver(keyword string, fn KeywordScopeResolveFunc) {
+	s.log.Debug("adding keyword resolution for '%v'", keyword)
+	s.keywordResolvers[keyword] = fn
+}
+
+func (s *ScopeResolver) AddAttributeResolver(prefix string, fn AttributeScopeResolveFunc) {
+	s.log.Debug("adding attribute resolution for '%v'", prefix)
+	s.attributeResolvers[prefix] = fn
+}
+
 func resolveUserSelf(u *models.SignedInUser) (string, error) {
 	return Scope("users", "id", fmt.Sprintf("%v", u.UserId)), nil
 }
 
-// ResolveKeyword resolves scope with keywords such as `self` or `current` into `id` based scopes
-func (s *ScopeResolver) ResolveKeyword(user *models.SignedInUser, permission Permission) (*Permission, error) {
-	if fn, ok := s.keywordResolvers[permission.Scope]; ok {
-		resolvedScope, err := fn(user)
+// GetResolveKeywordScopeMutator returns a function to resolve scope with keywords such as `self` or `current` into `id` based scopes
+func (s *ScopeResolver) GetResolveKeywordScopeMutator(user *models.SignedInUser) ScopeMutator {
+	return func(_ context.Context, scope string) (string, error) {
+		var err error
+		// By default the scope remains unchanged
+		resolvedScope := scope
+		if fn, ok := s.keywordResolvers[scope]; ok {
+			resolvedScope, err = fn(user)
+			if err != nil {
+				return "", fmt.Errorf("could not resolve %v: %w", scope, err)
+			}
+			s.log.Debug("resolved '%v' to '%v'", scope, resolvedScope)
+		}
+		return resolvedScope, nil
+	}
+}
+
+type AttributeScopeResolveFunc func(ctx context.Context, orgID int64, initialScope string) (string, error)
+
+// getCacheKey creates an identifier to fetch and store resolution of scopes in the cache
+func getCacheKey(orgID int64, scope string) string {
```

Comment:
Don't you also need the user?

Reply:
I'd say no, it's to translate a scope `datasources:name:_name_` into `datasources:id:_id_` whether the user can access it or not is evaluated after.

Reply:
And we're not going to cache for things like `users:me`?

But yeah, let's leave it out until we need it.